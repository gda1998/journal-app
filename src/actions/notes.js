import { db } from '../firebase/firebaseConfig';
import types from '../types/types';
import {
    addDoc,
    collection,
    doc,
    updateDoc,
    deleteDoc,
    // getDoc
} from 'firebase/firestore';
import Swal from 'sweetalert2';
import { swalAlert, swalLoading } from '../helpers/swalToast';
import loadNotes from '../helpers/loadNotes';
import fileUpload from '../helpers/fileUpload';
// import fileDelete from '../helpers/fileDelete';

export const startNewNoteAction = () => {
    return (dispatch, getState) => {
        const { uid } = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        const notesRef = collection(db, `${uid}/journal/notes`);
        return addDoc(notesRef, newNote)
            .then(docRef => {
                dispatch(activeNoteAction(docRef.id, newNote));
                dispatch(addNewNoteAction(docRef.id, newNote));
            })
            .catch(console.error);
    }
}

export const activeNoteAction = (id, note) => ({
    type: types.notesActive,
    payload: { id, ...note }
});

export const addNewNoteAction = (id, note) => ({
    type: types.notesAddNew,
    payload: { id, ...note }
});

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotesAction(notes));
    }
}

export const setNotesAction = (notes) => ({
    type: types.notesLoad,
    payload: notes
});

export const startUpdateNote = (note) => {
    return (dispatch, getState) => {
        const { uid } = getState().auth;
        !note.url && (note.url = null);
        const noteRef = doc(db, `${uid}/journal/notes/`, note.id);
        updateDoc(noteRef, note);
        dispatch(refreshNoteAction(note));
        swalAlert('Updated!', 'The note was updated', 'success');
    }
}

export const refreshNoteAction = (note) => ({
    type: types.notesUpdate,
    payload: note
});

export const startUploadingImgAction = (file) => {
    return async (dispatch, getState) => {
        const { active: activeNote } = getState().notes;
        swalLoading('Uploading', 'Please wait...');

        const fileUrl = await fileUpload(file);
        activeNote.url = fileUrl;

        Swal.close();
        await dispatch(startUpdateNote(activeNote));
    }
}

export const startDeleteNote = (id) => {
    return async (dispatch, getState) => {
        const { uid } = getState().auth;
        try {
            const ref = await doc(db, uid, `journal/notes/${id}`);
            // const noteRef = await getDoc(ref);

            // Delete the image from cloudinary and from the db
            // fileDelete(noteRef.data().url)
            await deleteDoc(ref);

            dispatch(deleteNote(id));
        } catch (error) {
            console.error('Error al eliminar la nota: ', error);
        }
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
});

export const notesLogout = () => ({
    type: types.notesLogoutCleaning
})