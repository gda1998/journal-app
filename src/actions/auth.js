import { 
    createUserWithEmailAndPassword,
    getAuth,
    signInWithPopup,
    signInWithEmailAndPassword,
    signOut,
    updateProfile 
} from 'firebase/auth';
import { googleAuthProvider } from '../firebase/firebaseConfig';
import { finishLoadingAction, startLoadingAction } from './ui';
import { swalAlert } from '../helpers/swalToast';
import types from '../types/types';
import { notesLogout } from './notes';

export const startRegisterAction = ({ email, password, name }) => {
    return (dispatch) => {
        const auth = getAuth();
        dispatch( startLoadingAction() );
        createUserWithEmailAndPassword(auth, email, password)
        .then( async({ user }) => {
            await updateProfile(user, { displayName: name });
            dispatch( login(user.uid, user.displayName) );
        })
        .catch( e => swalAlert('Error!', e, 'error') )
        .finally( () => dispatch( finishLoadingAction() ) );
    };
};

export const startLoginEmailPassword = ({ email, password }) => {
    return (dispatch) => {
        const auth = getAuth();
        dispatch( startLoadingAction() );
        return signInWithEmailAndPassword(auth, email, password)
        .then( ({user}) => dispatch( login(user.uid, user.displayName) ) )
        .catch( e => swalAlert('Error!', e, 'error') )
        .finally( () => dispatch( finishLoadingAction() ) );
    };
};

export const startGoogleLogin = () => {
    return (dispatch) => {
        const auth = getAuth();
        signInWithPopup(auth, googleAuthProvider)
        .then( ({user}) => {
            dispatch( login(user.uid, user.displayName) )
        });
    }
}

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
})

export const startLogout = () => {
    return async(dispatch) => {
        await signOut( getAuth() );
        dispatch( logout() );
        dispatch(  notesLogout() );
    }
}