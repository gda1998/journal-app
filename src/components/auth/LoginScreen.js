import React from 'react';
import { useDispatch } from 'react-redux';
import { Link } from 'react-router-dom';
import { login } from '../../actions/auth';
import { useForm } from '../../hooks/useForm';

export const LoginScreen = () => {

    const dispatch = useDispatch();

    const [ formValues, handleInputChange ] = useForm({
        email: 'gabriel@gmail.com',
        password: '123456'
    });

    const { email, password } = formValues;

    const handleLogin = e => {
        e.preventDefault();
        dispatch( login(1234, 'Hernando') );   
    }

    return (
        <>
            <h3 className="auth__title">Login</h3>
            <form onSubmit={ handleLogin }>

                <input
                    autoComplete="off"
                    className="auth__input"
                    type="email"
                    placeholder="Email"
                    name="email"
                    value={ email }
                    onChange={ handleInputChange }
                />
                <input
                    className="auth__input"
                    type="password"
                    placeholder="Password"
                    name="password"
                    value={ password }
                    onChange={ handleInputChange }
                />
                <button type="submit" className="btn btn-primary btn-block">
                    Login
                </button>

                <div className="auth__social-networks">
                    <p>Login with social networks</p>
                    <div className="google-btn">
                        <div className="google-icon-wrapper">
                            <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
                        </div>
                        <p className="btn-text">
                            <b>Sign in with google</b>
                        </p>
                    </div>{/* /.google-btn */}
                </div>{/* /.auth__social-networks */}

                <Link to="/auth/register" className="link">
                    Create a new account
                </Link>

            </form>
        </>
    );
}
