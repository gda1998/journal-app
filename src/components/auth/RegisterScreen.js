import { Link } from 'react-router-dom';
import { useDispatch, useSelector } from 'react-redux';
import validator from 'validator';
import { useForm } from '../../hooks/useForm';
import { startRegisterAction } from '../../actions/auth';
import { removeErrorAction, setErrorAction } from '../../actions/ui';

export const RegisterScreen = () => {
    const dispatch = useDispatch();
    const { msgError, loading } = useSelector( state => state.ui );
    const [ formValues, handleInputChange ] = useForm({
        name: '',
        email: '',
        password: '',
        password2: ''
    });
    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        isFormValid() && dispatch( startRegisterAction(formValues) );
    }

    const isFormValid = () => {
        if( !validator.isEmail(email) ){
            dispatch( setErrorAction('Email Invalid') );
            return false;
        }
        if( password !== password2 || password.length < 6 ){
            dispatch( setErrorAction('Password should be at least 6 characters and match each other') );
            return false;
        }
        dispatch( removeErrorAction() );
        return true;
    }

    return (
        <>
            <h3 className="auth__title">Register</h3>
            <form
                className="animate__animated animate__fadeIn animate__faster"
                onSubmit={ handleRegister }
            >
                {
                    msgError && 
                    <div className="auth__alert-error">
                        { msgError }
                    </div>
                }

                <input
                    autoComplete="off"
                    className="auth__input"
                    maxLength={255}
                    name="name"
                    onChange={ handleInputChange }
                    pattern="[a-zA-ZÀ-ÿ\u00f1\u00d1]+(\s*[a-zA-ZÀ-ÿ\u00f1\u00d1]*)*"
                    placeholder="Name"
                    required
                    title="El campo debe tener solamente caracteres alfabéticos."
                    type="text"
                    value={ name }
                />

                <input
                    autoComplete="off"
                    className="auth__input"
                    maxLength={255}
                    name="email"
                    onChange={ handleInputChange }
                    pattern="[a-zA-Z0-9_]+([.\-][a-zA-Z0-9_]+)*@[a-zA-Z0-9_]+([.][a-zA-Z0-9_]+)*[.][a-zA-Z]{1,5}"
                    placeholder="Email"
                    required
                    title="ej. example@email.com"
                    type="email"
                    value={ email }
                />
                <input
                    className="auth__input"
                    minLength={ 6 }
                    maxLength={255}
                    name="password"
                    onChange={ handleInputChange }
                    placeholder="Password"
                    required
                    type="password"
                    value={ password }
                />
                <input
                    className="auth__input"
                    minLength={ 6 }
                    maxLength={255}
                    name="password2"
                    onChange={ handleInputChange }
                    placeholder="Confirm"
                    required
                    type="password"
                    value={ password2 }
                />
                <button 
                    className="btn btn-primary btn-block mb-5"
                    disabled={ loading }
                    type="submit" 
                >
                    Register
                </button>

                <Link to="/auth/login" className="link">
                    Already registered?
                </Link>

            </form>
        </>
    );
}