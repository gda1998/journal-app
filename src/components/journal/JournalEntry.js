import dayjs from 'dayjs';
import { useDispatch } from 'react-redux';
import { activeNoteAction } from '../../actions/notes';


export const JournalEntry = ({ id, title, body, date, url }) => {
    const noteDate = dayjs(date);
    const dispatch = useDispatch();
    const handleEntryClick = () => dispatch( activeNoteAction(id, { title, body, date, url }) );

    return (
        <div 
            className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
            onClick={ handleEntryClick }
        >

            { url && 
                <div 
                    className="journal__entry-picture"
                    style={{ 
                        backgroundSize: 'cover',
                        backgroundImage: `url(${ url })`
                    }}
                >
                </div>/* /.journal__entry-picture */
            }


            <div className="journal__entry-body">
                <p className="journal__entry-title">
                    { title }
                </p>
                <p className="journal__entry-content">
                    { body }
                </p>
            </div>{/* /.journal__entry-body */}

            <div className="journal__entry-date-box">
                <span>{ noteDate.format('dddd') }</span>
                <h4>{ noteDate.format('DD') }</h4>
            </div>{/* /.journal__entry-date-box */}

        </div>/* /.journal__entry */
    );
}