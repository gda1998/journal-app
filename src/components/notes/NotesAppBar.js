import { useDispatch, useSelector } from 'react-redux';
import { startUpdateNote, startUploadingImgAction } from '../../actions/notes';

export const NotesAppBar = () => {
    // Redux
    const dispatch = useDispatch();
    const { active:note } = useSelector( state => state.notes );
    const handleUpdate = () => dispatch( startUpdateNote(note) );
    const handlePictureClick = () => document.querySelector('#fileSelector').click();
    
    const handleFileChange = ({target}) => {
        const file = target.files[0];
        file && dispatch( startUploadingImgAction(file) );
    }

    return (
        <div className="notes__appbar">
            <span>05 de Noviembre de 2021</span>

            <input
                id="fileSelector"
                onChange ={ handleFileChange }
                type="file"
                style={{ display: 'none' }}
            />

            <div>
                <button 
                    className="btn"
                    onClick={ handlePictureClick }
                >
                    Picture
                </button>
                <button 
                    className="btn"
                    onClick={ handleUpdate }
                >
                    Save
                </button>
            </div>
        </div>/* /.notes__appbar */
    );
}