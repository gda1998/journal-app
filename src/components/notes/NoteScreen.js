import { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useForm } from '../../hooks/useForm';
import { NotesAppBar } from './NotesAppBar';
import { activeNoteAction, startDeleteNote } from '../../actions/notes';

export const NoteScreen = () => {
    // Redux
    const dispatch = useDispatch();
    const { active:note } = useSelector( state => state.notes );

    // useForm
    const [ formValues, handleInputChange, reset ] = useForm(note);
    const { title, body } = formValues;
    note.url && ( formValues.url = note.url );

    // useRef
    const activeId = useRef(note.id);

    // handleDelete
    const handleDelete = () => dispatch( startDeleteNote(note.id) );

    useEffect(() => {
        if( activeId.current !== note.id ){
            reset(note);
            activeId.current = note.id;
        }
    }, [note, reset]);

    useEffect(() => {
        dispatch( activeNoteAction(formValues.id, {...formValues}) );
    }, [dispatch, formValues]);
    

    return (
        <div className="notes__main-content">

            <NotesAppBar />
            <div className="notes__content">

                <input
                    autoComplete="off"
                    className="notes__title-input"
                    name="title"
                    onChange={ handleInputChange }
                    placeholder="Some awesome title"
                    type="text"
                    value={ title }
                />
                <textarea
                    className="notes__textarea"
                    name="body"
                    onChange={ handleInputChange }
                    placeholder="What's happened today?"
                    type="text"
                    value={ body }
                ></textarea>

                {
                    note.url &&
                    <div className="notes__image">
                        <img
                            src={ note.url }
                            alt="imagen"
                        />
                    </div>/* /.notes__image */
                }

            </div>{/* /.notes__content */}

            <button
                className="btn btn-danger"
                onClick={ handleDelete }
            >
                Delete
            </button>

        </div>/* /.notes__main-content */
    );
}
