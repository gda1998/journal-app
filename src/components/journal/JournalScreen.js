import { useSelector } from 'react-redux';
import { Navigate } from 'react-router-dom';
import { NoteScreen } from '../notes/NoteScreen';
import { Sidebar } from './Sidebar';
import { NothingSelected } from './NothingSelected';

export const JournalScreen = () => {
    const { auth, notes } = useSelector( state => state );
    return Object.keys(auth).length === 0 
    ? <Navigate to="/auth/login" />
    : (
        <div className="journal__main-content animate__animated animate__fadeIn animate__faster">
            <Sidebar />
            <main>
                {
                    notes.active === null ? <NothingSelected /> : <NoteScreen />
                }
            </main>
        </div>
    );
}
