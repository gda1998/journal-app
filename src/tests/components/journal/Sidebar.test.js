import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { startLogout } from '../../../actions/auth';
import { startNewNoteAction } from '../../../actions/notes';
import { Sidebar } from '../../../components/journal/Sidebar';

jest.mock('../../../actions/auth', () => ({
    ...jest.requireActual(),
    startLogout: jest.fn()
}));

jest.mock('../../../actions/notes', () => ({
    ...jest.requireActual(),
    startNewNoteAction: jest.fn()
}));

const mockStore = configureStore([thunk]);
const initState = {
    ui: {},
    auth: { uid: '1234', name: 'testing' },
    notes: {
        active: {},
        notes: [{ id:'11111', date: 1647071655611 }, { id: '22222',date: 1647071655611 }]
    }
};
let store = mockStore(initState);

describe('Pruebas en <Sidebar />', () => {
    let wrapper;
    const getWrapper = (store) => mount(
        <Provider store={store}>
            <Sidebar />
        </Provider>
    );

    beforeEach(() => {
        wrapper = getWrapper(store);
        jest.clearAllMocks();
        store.clearActions();
    })

    test('Debe de mostrarse correctamente', () => { 
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe de mostrarse el nombre del usuario', () => {
        const spanText = wrapper.find('span').at(0).text();
        expect(spanText).toBe('testing');
    });
    
    test('Debe de hacer el logout del sistema', () => {
        const store = mockStore(initState);
        store.dispatch = jest.fn();
        const wrapper = getWrapper(store);

        wrapper.find('button').at(0).simulate('click');
        expect(startLogout).toHaveBeenCalled();
    });

    test('Debe de llamar a startNewNoteAction para crear una nota', () => {
        const store = mockStore(initState);
        store.dispatch = jest.fn();
        const wrapper = getWrapper(store);

        wrapper.find('.journal__new-entry').simulate('click');
        expect(startNewNoteAction).toHaveBeenCalled();
    });
});