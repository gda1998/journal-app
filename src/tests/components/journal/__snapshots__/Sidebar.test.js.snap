// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <Sidebar /> Debe de mostrarse correctamente 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar"
  >
    <h3
      className="mt-5"
    >
      <i
        className="far fa-moon"
      />
      <span>
        testing
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-5x"
    />
    <p
      className="mt-5"
    >
      New Entry
    </p>
  </div>
  <div
    className="journal__entries"
  >
    <div
      className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
      onClick={[Function]}
    >
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        />
        <p
          className="journal__entry-content"
        />
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
          Saturday
        </span>
        <h4>
          12
        </h4>
      </div>
    </div>
    <div
      className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
      onClick={[Function]}
    >
      <div
        className="journal__entry-body"
      >
        <p
          className="journal__entry-title"
        />
        <p
          className="journal__entry-content"
        />
      </div>
      <div
        className="journal__entry-date-box"
      >
        <span>
          Saturday
        </span>
        <h4>
          12
        </h4>
      </div>
    </div>
  </div>
</aside>
`;
